File: PythonTests/ComputationGraph.py
from MikeLearn import Node
from MikeLearn import ComputationGraph

comp_graph = ComputationGraph()

n = list();
for i in range(0,3):
    n.append(comp_graph.AddNode())

n.append(comp_graph.AddZNode())
comp_graph.setParent(0,3)
comp_graph.setParent(1,3)
comp_graph.setParent(2,3)
comp_graph.setLeaf(1)
comp_graph.setLeaf(2)

comp_graph.setRootNode(3)

#xor

x0 = [1,0]
nCol = 1
comp_graph.setNodeData(0,x0,nCol)

W1 = [0.5,0.4,0.3,0.1]
nCol = 2
comp_graph.setNodeData(1,W1,nCol)

B1 = [0.1,0.1]
nCol = 1
comp_graph.setNodeData(2,B1,nCol)

comp_graph.printstuff();

File: MikeLearn/ComputationGraph.cpp
#include "ComputationGraph.h"
#include "ZNode.h"
int ComputationGraph::AddNode()
{
	Node* newNode = new Node();
	newNode->setId(currentNodeIndex);
	NodeVector.push_back(newNode);
	currentNodeIndex++;

	return currentNodeIndex - 1;
}

int ComputationGraph::AddZNode()
{
	ZNode* newNode = new ZNode();
	newNode->setId(currentNodeIndex);
	NodeVector.push_back(newNode);
	currentNodeIndex++;

	return currentNodeIndex - 1;
}

void ComputationGraph::setParent(int childId, int parentId)
{
	NodeVector[childId]->setParent(NodeVector[parentId]);
}

void ComputationGraph::setLeaf(int leafNode)
{

	NodeVector[leafNode]->setIsLeaf(true);

}

void ComputationGraph::printstuff()
{
	Logger logger = Logger(1);
	for (Node* node : NodeVector)
	{
		logger.LogGeneral("========", "Node",node->getID(), "========");
		for (Node* child : node->getChildren())
		{
			logger.LogGeneral(child->getID());

		}
	}

	std::vector<Node*> path;
	setLossPath(rootNode, path);
	
	for (Node* node : NodeVector)
	{
		if (node->getIsLeaf())
		{
			logger.LogGeneral("========", "Node", node->getID(), "========");
			node->printPathToLoss();
		}


	}

	forward(NodeVector[0]);

}

void ComputationGraph::setLossPath(Node* node, std::vector<Node*> path)
{
	path.push_back(node);

	if (node->getIsLeaf() == true)
	{
		node->setPathToLoss(path);
	}

	for (Node* child : node->getChildren())
	{

		setLossPath(child, path);

	}

}

void ComputationGraph::setNodeData(int nodeId, std::vector<float> inData, int nCol)
{
	NodeVector[nodeId]->setData(inData, nCol);
}

void ComputationGraph::forward(Node* node)
{
	Logger logger = Logger(1);
	ZNode* zNode = dynamic_cast<ZNode*>(node);
	logger.LogGeneral("==============", "Node", node->getID(), "==============");
	if (zNode != nullptr)
	{
		logger.LogGeneral("Cast succesful!");
		zNode->Compute();
	}
	else
	{
		logger.LogGeneral("Cast unsuccesful!");
	}

	if (node == rootNode)
	{
		return;
	}
	else
	{
		forward(node->getParent());
	}


}
